Runtime/Services/BacktraceSession.cs

```diff
@@ -0,0 +1,317 @@
+        /// <param name="maximumNumberOfEventsInStore">Determine how many events we can store in event store</param>
+        public BacktraceSession(
+            AttributeProvider attributeProvider,
+            string uploadUrl,
+            long timeIntervalInMs,
+            uint maximumNumberOfEventsInStore)
+        {
+            SubmissionUrl = uploadUrl;
+            _attributeProvider = attributeProvider;
+            _maximumNumberOfEventsInStore = maximumNumberOfEventsInStore;
+            _timeIntervalInMs = timeIntervalInMs;
+            RequestHandler = new BacktraceHttpClient();
+        }
+
+        /// <summary>
+        /// Backtrace session update interval method used by Backtrace client to update session events time
+        /// </summary>
+        /// <param name="time">Current game time</param>
+        public void Tick(long time)
+        {
+            if (_timeIntervalInMs == 0)
+            {
+                return;
+            }
+            lock (_object)
+            {
+                var intervalUpdate = (time - _lastUpdateTime) >= _timeIntervalInMs;
+                var reachedEventLimit = _maximumNumberOfEventsInStore == Count() && _maximumNumberOfEventsInStore != 0;
+                if (intervalUpdate == false && reachedEventLimit == false)
+                {
+                    // nothing more to update
+                    return;
+                }
+                _lastUpdateTime = time;
+            }
+            Send();
+        }
+
+        /// <summary>
+        /// Force BacktraceSession to Send data to Backtrace
+        /// </summary>
+        public void Send()
+        {
+            Send(numberOfRetries: 3);
+        }
+
+        /// <summary>
+        /// Force BacktraceSession to Send data to Backtrace with retry settings
+        /// </summary>
+        public void Send(int numberOfRetries = DefaultNumberOfRetries)
+        {
+            if (numberOfRetries == 0)
+            {
+                Debug.LogWarning("Backtrace Session: Cannot send session data to Backtrace due to submission issue.");
+                CleanEvents();
+                return;
+            }
+            if (Count() == 0)
+            {
+                return;
+            }
+            // create a copy of events 
+            var uniqueEvents = UniqueEvents.ToArray();
+            var sessionEvents = SessionEvents.ToArray();
+            var payload = CreateJsonPayload(uniqueEvents, sessionEvents);
+
+            // cleanup existing copy of events
+            CleanEvents();
+
+            // submit data to Backtrace
+            RequestHandler.Post(SubmissionUrl, payload, (long statusCode, bool httpError, string response) =>
+            {
+                if (statusCode == 200)
+                {
+                    OnRequestCompleted();
+                }
+                else
+                {
+                    OnRequestFailure(uniqueEvents, sessionEvents);

```

Comment:
because `OnrRequestFailure` always re-adds these events, doesn't this mean that we effectively retry sending them forever while blocking any further events the user tries to add?

Reply:
You're right. We can make this logic simpler. I will expose send and retry logic to another function to make it cleaner for other developers to read this code. Thanks!